webapp/src/actions.ts

```diff
@@ -295,3 +297,83 @@ export const displayGenericErrorModal = (title: MessageDescriptor, message: Mess
         return {};
     };
 };
+
+export function incomingCallOnChannel(channelID: string, hostID: string, startAt: number) {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
+        let channel = getChannel(getState(), channelID);
+        if (!channel) {
+            const res = await dispatch(loadChannel(channelID));
+            channel = res.data;
+        }
+
+        if (channel && (isDMChannel(channel) || isGMChannel(channel))) {
+            if (voiceChannelCallDismissedNotification(getState(), channelID)) {
+                return;
+            }
+
+            const otherUser = getUser(getState(), hostID);
+            if (!otherUser) {
+                await dispatch(getProfilesByIdsAction([hostID]));
+            }
+
+            // if this is the first call in the list, it should ring.
+            const calls = incomingCalls(getState());
+            const ring = calls.length === 0;
+
+            await dispatch({
+                type: ADD_INCOMING_CALL,
+                data: {
+                    callID: channelID,
+                    hostID,
+                    startAt,
+                    type: isDMChannel(channel) ? ChannelType.DM : ChannelType.GM,
+                    ring,
+                },
+            });
+        }
+    };
+}
+
+export const userDisconnected = (channelID: string, userID: string) => {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
+        await dispatch({
+            type: VOICE_CHANNEL_USER_DISCONNECTED,
+            data: {
+                channelID,
+                userID,
+                currentUserID: getCurrentUserId(getState()),
+            },
+        });
+
+        if (!channelHasCall(getState(), channelID)) {
+            await dispatch({
+                type: CALL_HAS_ENDED,
+                data: {
+                    callID: channelID,
+                },
+            });
+        }
+    };
+};
+
+export const dismissIncomingCallNotification = (callID: string) => {
+    return async (dispatch: DispatchFunc) => {
+        Client4.doFetch(
+            `${getPluginPath()}/calls/${callID}/dismiss-notification`,
+            {method: 'post'},
+        );
+        await dispatch(removeIncomingCallNotification(callID));
```

Comment:
Good question. Nope, I specifically didn't wait on it so that making the notification and ringing go away wasn't delayed. I added a catch just to clean up a bit better.